File: json-tools/scripts/stats.py
#!/usr/bin/env python3
# Copyright 2026 Yevgeniy Zaremba
# SPDX-License-Identifier: Apache-2.0

"""Generate summary statistics for JSON records."""

from __future__ import annotations

import argparse
import math
from collections import Counter, defaultdict
from typing import Any

from common import extract_values, frequency, load_json, resolve_array, type_name, unique_types, write_json


def parse_fields(raw: str | None) -> list[str]:
    if not raw:
        return []
    return [item.strip() for item in raw.split(",") if item.strip()]


def numeric_summary(values: list[Any]) -> dict[str, Any]:
    nums = [float(v) for v in values if isinstance(v, (int, float)) and not isinstance(v, bool)]
    if not nums:
        return {}
    return {
        "count": len(nums),
        "min": min(nums),
        "max": max(nums),
        "mean": sum(nums) / len(nums),
    }


def main() -> None:
    parser = argparse.ArgumentParser(description="Compute practical JSON stats for arrays of records.")
    parser.add_argument("input", nargs="?", default="-", help="Input JSON file path or '-' for stdin.")
    parser.add_argument("--array-path", help="Path to the array to analyze.")
    parser.add_argument("--fields", help="Comma-separated field paths to analyze.")
    parser.add_argument("--top", type=int, default=10, help="Top N frequent values to include.")
    parser.add_argument("--compact", action="store_true", help="Emit compact JSON output.")
    args = parser.parse_args()

    data = load_json(args.input)
    records = resolve_array(data, args.array_path)
    if not records:
        records = data if isinstance(data, list) else [data]

    selected_fields = parse_fields(args.fields)
    if not selected_fields:
        field_candidates: set[str] = set()
        for record in records:
            if isinstance(record, dict):
                field_candidates.update(record.keys())
        selected_fields = sorted(field_candidates)

    field_stats: dict[str, Any] = {}
    for field in selected_fields:
        values: list[Any] = []
        presence = 0
        for record in records:
            found = extract_values(record, field)
            if found:
                presence += 1
                values.extend(found)
        freq = frequency(values)
        has_complex = any(isinstance(v, (dict, list)) for v in values)
        entry: dict[str, Any] = {
            "presence": f"{presence}/{len(records)}",
            "types": unique_types(values),
            "unique_values": len(freq),
        }
        if not has_complex:
            entry["top_values"] = [
                {"value": key, "count": count}
                for key, count in freq.most_common(max(args.top, 0))
            ]
        field_stats[field] = entry
        numeric = numeric_summary(values)
        if numeric:
            field_stats[field]["numeric"] = numeric

    result = {
        "record_count": len(records),
        "field_count": len(selected_fields),
        "fields": field_stats,
    }
    write_json(result, compact=args.compact)


if __name__ == "__main__":
    main()

File: json-tools/scripts-node/stats.mjs
#!/usr/bin/env node
// Copyright 2026 Yevgeniy Zaremba
// SPDX-License-Identifier: Apache-2.0

import { parseArgs } from "node:util";
import { extractValues, frequency, loadJson, resolveArray, uniqueTypes, writeJson } from "./common.mjs";

function parseFields(raw) {
  if (!raw) {
    return [];
  }
  return raw.split(",").map((item) => item.trim()).filter(Boolean);
}

function numericSummary(values) {
  const nums = values.filter((v) => typeof v === "number" && !Number.isNaN(v));
  if (!nums.length) {
    return {};
  }
  return {
    count: nums.length,
    min: Math.min(...nums),
    max: Math.max(...nums),
    mean: nums.reduce((acc, item) => acc + item, 0) / nums.length,
  };
}

function main() {
  const { values, positionals } = parseArgs({
    options: {
      "array-path": { type: "string" },
      fields: { type: "string" },
      top: { type: "string", default: "10" },
      compact: { type: "boolean", default: false },
    },
    allowPositionals: true,
  });

  const input = positionals[0] ?? "-";
  const top = Number.parseInt(values.top, 10);
  const data = loadJson(input);
  let records = resolveArray(data, values["array-path"]);
  if (!records.length) {
    records = Array.isArray(data) ? data : [data];
  }

  let selectedFields = parseFields(values.fields);
  if (!selectedFields.length) {
    const fieldCandidates = new Set();
    for (const record of records) {
      if (record && typeof record === "object" && !Array.isArray(record)) {
        for (const key of Object.keys(record)) {
          fieldCandidates.add(key);
        }
      }
    }
    selectedFields = [...fieldCandidates].sort();
  }

  const fieldStats = {};
  for (const field of selectedFields) {
    const fieldValues = [];
    let presence = 0;
    for (const record of records) {
      const found = extractValues(record, field);
      if (found.length) {
        presence += 1;
        fieldValues.push(...found);
      }
    }
    const freq = frequency(fieldValues);
    const hasComplex = fieldValues.some((v) => v && typeof v === "object");
    const entry = {
      presence: `${presence}/${records.length}`,
      types: uniqueTypes(fieldValues),
      unique_values: freq.size,
    };
    if (!hasComplex) {
      const topValues = [...freq.entries()]
        .sort((a, b) => b[1] - a[1])
        .slice(0, Math.max(top, 0))
        .map(([value, count]) => ({ value, count }));
      entry.top_values = topValues;
    }
    const numeric = numericSummary(fieldValues);
    if (Object.keys(numeric).length) {
      entry.numeric = numeric;
    }
    fieldStats[field] = entry;
  }

  writeJson({
    record_count: records.length,
    field_count: selectedFields.length,
    fields: fieldStats,
  }, values.compact);
}

main();
